File: java/src/learning/Algorithms.java
/* algorithms
    - Algorithms
        - are used to solve problems related to sorting, searching, or manipulating data strutures
            - Searching Algorithms
                - these types of algorithms are used to find certain elements inside a collection
                - a well-known searching algorithm is called "Binary Search"
            - Sorting Algoriths
                - these types of alrighthms are uses to sort collections
                    - depening on what order the resulting collection becomes
                - a common order type is natural order
            - Iteraion
                - iteration is also another type of algorithm
                    - used to loop through elements
                - refer to Iteratives.java for more info

    - Java Algorithms
        - Seaching
            - Java has various built-in helper methods for searching through collections
            - Binary Search
                - a commonly used searching algorithm
                - can be used through the .binarySearch() method inside the Collections class
                    - returns the index of the element if found
                - note that the collectrion must be sorted first
                - syntax:
                    Collections.binarySearch(<collection.Object>, <element>) 
            - Maximum and Minimum Search
                - used to find the maximum and minimum values of a collection
                    - through the .max() and .min() methods of the Collections class, respectively
                - syntax:
                    Collections.max(<collectionObject>);
                    // or
                    Collections.min(<collectionObject>);
            - Frequency Search
                - used to query a collection about the number of times an element appears
                    - through the .frequency() method of the Collections class
                - syntax:
                    Collections.frequency(<collectionObject>, <element>);
            - Disjoint Search
                - compares two collections is they have no elements in common
                - results into a boolean value
                    - returns true if there are no common elements between the two collections
                    - returns a false is there is
                - syntax:
                    Collections.disjoint(<collectionObject1>, <collectionObject2>);    

        - Sorting
            - Java has a built-in sorting method for sorting methods
            - Natural Order
                - a common type of sorting algorithm
                - sorts items alphanumerically
                    - 0-9, then
                    - A-Z, and
                    - a-z
                - can be used through the .sort() method inside the Collections class
                    - sorts the items through natural order
                - syntax:
                    Collections.sort(<collectionObject>);
            - Reverse Order
                - reverses the current order of a collection
                - used through the .reverse() method inside the Collections class
                - syntax:
                    Collections.reverse(<collectionObject>);
                - another method, reverseOrder() can be used inside the .sort() method
                    - syntax:
                        Collections.sort(<collectionObject>, reverseOrder());
            - Random Order
                - randomly shuffles the elements around in a collection
                    - used through the .shuffly() method from the Collections class
                - syntax:
                    Collections.shuffle(<collectionObject>);
            - Swap Order
                - swaps two elements inside a collection
                    - used through the .swap() method from the Collections class
                - syntax:
                    Collections.swap(<collectionObject>, <index1>, <index2>);
*/

import java.util.Collections;
import java.util.ArrayList;

public class Algorithms {
    static int fibonacciRecursion(int a) {
        if(a == 1 || a == 0) {
            return 1;
        }
        return fibonacciRecursion(a - 1) + fibonacciRecursion(a - 2);
    }
    
    public static void main(String[] args) {

    }
}

File: cpp/src/learning/algorithms.cpp
/* algorithms
    - Algorithms
        - are used to solve problems related to sorting, searching, or manipulating data strutures
            - Searching Algorithms
                - these types of algorithms are used to find certain elements inside a collection
                - a well-known searching algorithm is called "Binary Search"
            - Sorting Algoriths
                - these types of algorithms are uses to sort collections
                    - depending on what order the resulting collection becomes
                - a common order type is natural order
            - Iteration
                - iteration is also another type of algorithm
                    - used to loop through elements
                - refer to Iteratives.java for more info

    - C++ Algorithms
        - has various functions for searching or sorting data structures in C++
            - imported through the <algorithms> header library
            - syntax:
                #include <algorithms>
        - Sorting
            - std::sort()
                - sorts elements in a data structure
                    - sorting via lexicographical order (ASCII-based)
                - takes two arguments
                    - <start_iterator> is the starting point of a data structure
                    - <end_iterator> is the end point of a data structure
                - syntax:
                    std::sort(<start_iterator>, <end_iterator>);
                - example:
                    std::sort(myVector.begin(), myVector.end());
                - to reverse the order, the .rbegin() and .rend() methods are used
                    - example:
                        std::sort(myVector.rbegin(), myVector.rend());
        - Searching
            - std::find()
                - searches for specific elements in a data structure
                - takes three arguments:
                    - <start_iterator> is the starting point in a data structure
                    - <end_iterator> is the end point in a data structure
                    - <value> is the search item
                - syntax:
                    std::find(<start_iterator>, <end_iterator>, <value>);
                - example:
                    std::find(myDeque.begin(), myDeque.end(), 2);
            - std::upper_bound()
                - searches for the first element that is greater than the specified value
                    - returns the index of the first element that is greater than it
                - typically used in a sorted data structure
                - takes three arguments
                    - <start_iterator> is the starting point in a data structure
                    - <end_iterator> is the end point in a data structure
                    - <value> is the search item
                - syntax:
                    std::upper_bound(<start_iterator>, <end_iterator>, <value>);
                - examle:
                    std::upper_bound(myList.begin(), myList.end(), 1);
            - std::lower_bound()
                - searches for the first element that is greater than or equal to the specified value
                    - returns the index of the first element that is greater than or equal to it
                - typically used in a sorted data structure
                - takes three arguments
                    - <start_iterator> is the starting point in a data structure
                    - <end_iterator> is the end point in a data structure
                    - <value> is the search item
                - syntax:
                    std::lower_bound(<start_iterator>, <end_iterator>, <value>);
                - examle:
                    std::lower_bound(myList.begin(), myList.end(), 1);
            - std::min_element()
                - searches for the smallest element in a data structure
                - takes two arguments
                    - <start_iterator> is the starting point of a data structure
                    - <end_iterator> is the end point of a data structure
                - syntax:
                    std::min_element(<start_iterator>, <end_iterator>);
                - example:
                    std::min_element(myVector.begin(), myVector.end());
            - std::max_element()
                - searches for the largest element in a data structure
                - takes two arguments
                    - <start_iterator> is the starting point of a data structure
                    - <end_iterator> is the end point of a data structure
                - syntax:
                    std::max_element(<start_iterator>, <end_iterator>);
                - example:
                    std::max_element(myVector.begin(), myVector.end());
        - Modifying
            - std::copy()
                - copies elements from one data structure to the other
                - takes three arguments
                    - <source_start_iterator> is the starting point of the source data structure
                    - <soruce_end_iterator> is the end point of the source data structure
                    - <destination_start_iterator> is the starting point of the destination data structuer
                - syntax:
                    std::sort(<source_start_iterator>, <source_end_iterator>, <destination_start_iterator>);
                - example:
                    // if the destination is empty
                    std::copy(std::begin(myVector), std::end(myVector), std::begin(copyVector));

                    // if the destination is not empty or already has elements inside it
                    std::copy(std::begin(myVector), std::end(myVector), srd::back_inserter(copyVector)); // redirects new elements to .push_back()
            - std::fill()
                - fills a data structure with a specified value
                - takes three arguments
                    - <start_iterator> is the starting point of a data structure
                    - <end_iterator> is the end point of a data structure
                    - <value> is the fill value
                - syntax:
                    std::fill(<start_iterator>, <end_iterator>, <value>);
                - example:
                    std::fill(myVector.begin(), myVector.end(), 12);
*/